### PR TITLE
Move Boolean, String, Char & Number instances to typeclass companion

### PR DESCRIPTION
**Pull request by nomisRev**
This PR moves the instances of Boolean, String, Char & Number to the companions of their corresponding typeclasses, and adds deprecations for their currently public APIs.

### CODE REVIEW

**Review by raulraja — APPROVED** [review]
thanks @nomisRev !

File: arrow-core/src/main/kotlin/arrow/core/extensions/string.kt [hunk 1]
```diff
@@ -38,29 +60,50 @@ interface StringShow : Show<String> {
...
 
   override fun String.compareTo(b: String): Int = this.compareTo(b)
 }
 
+private object StringOrderInstance : StringOrder
+
+fun Order.Companion.String(): Order<String> =
+  StringOrderInstance
+
+@Deprecated("Typeclass instance have been moved to the companion object of the typeclass", ReplaceWith("Show.string()", "arrow.core.Show", "arrow.core.string"))
```
Comment (franciscodr) [inline arrow-core/src/main/kotlin/arrow/core/extensions/string.kt:84]:
```suggestion
@Deprecated("Typeclass instance have been moved to the companion object of the typeclass", ReplaceWith("Order.string()", "arrow.core.Order", "arrow.core.string"))
```

File: arrow-core/src/main/kotlin/arrow/core/extensions/string.kt [hunk 2]
```diff
@@ -38,29 +60,50 @@ interface StringShow : Show<String> {
...
 fun String.Companion.order(): Order<String> =
   object : StringOrder {}
 
 interface StringHash : Hash<String>, StringEq {
   override fun String.hash(): Int = hashCode()
 }
 
+@Deprecated("Typeclass instance have been moved to the companion object of the typeclass", ReplaceWith("Show.string()", "arrow.core.Show", "arrow.core.string"))
```
Comment (franciscodr) [inline arrow-core/src/main/kotlin/arrow/core/extensions/string.kt:92]:
```suggestion
@Deprecated("Typeclass instance have been moved to the companion object of the typeclass", ReplaceWith("Hash.string()", "arrow.core.Hash", "arrow.core.string"))
```

**Comment by nomisRev** [comment]
Three times in a row @franciscodr. In Belgium we say "drie keer is trakteren", so I owe you a 🍺! Thanks 🕵️‍♂️.

**Review by franciscodr — APPROVED** [review]
@nomisRev I take up your offer 😄

**Pull request merged**